src/Interpreter.h

```diff
@@ -16,743 +16,128 @@
     int incCounter() {
         return counter++;
     }
 
-    /**
-     * Obtains a mutable reference to one of the relations maintained
-     * by this environment. If the addressed relation does not exist,
-     * a new, empty relation will be created.
-     */
-    InterpreterRelation& getRelation(const RamRelation& id) {
-        InterpreterRelation* res = nullptr;
+    /** Create relation */
+    void createRelation(const RamRelation& id) {
         auto pos = data.find(id.getName());
-        if (pos != data.end()) {
-            res = (pos->second);
+        InterpreterRelation* res = nullptr;
+        assert(pos == data.end());
+        if (!id.isEqRel()) {
+            res = new InterpreterRelation(id.getArity());
         } else {
-            if (!id.isEqRel()) {
-                res = new InterpreterRelation(id.getArity());
-            } else {
-                res = new InterpreterEqRelation(id.getArity());
-            }
-            data[id.getName()] = res;
+            res = new InterpreterEqRelation(id.getArity());
         }
-        // return result
-        return *res;
+        data[id.getName()] = res;
     }
 
-    /**
-     * Obtains an immutable reference to the relation identified by
-     * the given identifier. If no such relation exist, a reference
-     * to an empty relation will be returned (not exhibiting the proper
-     * id, but the correct content).
-     */
-    const InterpreterRelation& getRelation(const RamRelation& id) const {
+    /** Get relation */
+    InterpreterRelation& getRelation(const std::string& name) {
         // look up relation
-        auto pos = data.find(id.getName());
+        auto pos = data.find(name);
         assert(pos != data.end());
-
-        // cache result
         return *pos->second;
     }
 
-    /**
-     * Obtains an immutable reference to the relation identified by
-     * the given identifier. If no such relation exist, a reference
-     * to an empty relation will be returned (not exhibiting the proper
-     * id, but the correct content).
-     */
-    const InterpreterRelation& getRelation(const std::string& name) const {
-        auto pos = data.find(name);
-        assert(pos != data.end());
-        return *pos->second;
+    /** Get relation */
+    inline InterpreterRelation& getRelation(const RamRelation& id) {
+        return getRelation(id.getName());
     }
 
-    /**
-     * Returns the relation map
-     */
+    /** Get relation map */
     relation_map& getRelationMap() const {
         return const_cast<relation_map&>(data);
     }
 
-    /**
-     * Tests whether a relation with the given name is present.
-     */
+    /** Check for relation */
     bool hasRelation(const std::string& name) const {
         return data.find(name) != data.end();
     }
 
-    /**
-     * Deletes the referenced relation from this environment.
-     */
+    /** Drop relation */
     void dropRelation(const RamRelation& id) {
         data.erase(id.getName());
```

Comment:
The relation should be deleted before the entry is erased. We're currently leaking memory.

Reply:
Good point...

Reply:
It seems that we also sometimes drop relations that don't exist. I'm not sure if this is intended, and dropRelation should handle it, or unintended and there is a bug elsewhere.

Reply:
Oh, but it looks like you may have already fixed that.